pageserver/src/tenant_guard.rs

```diff
@@ -0,0 +1,138 @@
+    tenant: Option<Tenant>,
+    drop_watcher: watch::Sender<()>,
+}
+
+impl Drop for TenantGuardInner {
+    fn drop(&mut self) {
+        let tenant = self.tenant.take().unwrap();
+        let tenant_id = tenant.tenant_id();
+        let future = async move {
+            debug!("shutdown tenant {tenant_id}");
+
+            // Shut down all existing walreceiver connections and stop accepting the new ones.
+            task_mgr::shutdown_tasks(Some(TaskKind::WalReceiverManager), None, None).await;
+
+            // Ok, no background tasks running anymore. Flush any remaining data in
+            // memory to disk.
+            //
+            // We assume that any incoming connections that might request pages from
+            // the tenant have already been terminated by the caller, so there
+            // should be no more activity in any of the repositories.
+            //
+            // On error, log it but continue with the shutdown for other tenants.
+            if let Err(err) = tenant.checkpoint() {
+                error!("Could not checkpoint tenant {tenant_id} during shutdown: {err:?}");
+            }
+
+            Ok(())
+        };
+
+        spawn(
+            BACKGROUND_RUNTIME.handle(),
+            TaskKind::TenantDrop,
+            Some(tenant_id),
+            None,
+            "tenant drop",
+            false,
+            future,
+        );
+
+        self.drop_watcher.send_replace(());
+    }
+}
+
+#[derive(Clone)]
+pub struct TenantGuard(Weak<TenantGuardInner>);
+
+impl TenantGuard {
+    pub fn new(tenant: Tenant) -> Self {
+        let mut recv = tenant.subscribe_for_state_updates();
+        let (drop_watcher, _) = watch::channel(());
+        let loop_arc = Arc::new(TenantGuardInner {
+            tenant: Some(tenant),
+            drop_watcher,
+        });
+        let weak_inner = Arc::downgrade(&loop_arc);
+
+        let future = async move {
+            while recv.changed().await.is_ok() {
+                match *recv.borrow_and_update() {
+                    TenantState::Active { .. } | TenantState::Paused => continue,
+                    TenantState::Broken | TenantState::Stopping => break,
+                    TenantState::Stopped => unreachable!(),
+                }
+            }
+            drop(loop_arc)
+        };
+        BACKGROUND_RUNTIME.spawn(future);
+
+        Self(weak_inner)
+    }
+
+    pub fn current_state(&self) -> TenantState {
+        match self.0.upgrade() {
+            Some(inner) => inner.tenant.as_ref().unwrap().current_state(),
+            None => TenantState::Stopped,
+        }
+    }
+
+    pub fn is_active(&self) -> bool {
+        matches!(self.current_state(), TenantState::Active { .. })
+    }
+
+    pub fn subscribe_for_tenant_shutdown(&self) -> Option<watch::Receiver<()>> {
+        Some(self.0.upgrade()?.drop_watcher.subscribe())
+    }
+
+    pub fn with_upgrade<F, T>(&self, f: F) -> T
+    where
+        F: FnOnce(anyhow::Result<TenantAccessor>) -> T,
```

Comment:
Instead, I think it's better to return `Result<O>` from `with_upgrade` (changed to `with_tenant` from the last time). This shows whether the closure was executed, and I think that proved easy to use when I refactored the code later. What about this approach?

IMO there's one problem currently: we can't distinguish between "no tenant locally" and "tenant present but is dead" without checking the error itself. But looks like it will only add complexity to the code without benefits.